tendermint/src/abci/response.rs

```diff
@@ -0,0 +1,1027 @@
+    pub last_block_app_hash: Bytes,
+}
+
+/// Returned on genesis after initializing chain state.
+///
+/// [ABCI documentation](https://docs.tendermint.com/master/spec/abci/abci.html#initchain)
+#[derive(Clone, PartialEq, Eq, Debug, Default)]
+pub struct InitChain {
+    /// Initial consensus-critical parameters (optional).
+    pub consensus_params: Option<ConsensusParams>,
+    /// Initial validator set (optional).
+    ///
+    /// If this list is empty, the initial validator set will be the one given in
+    /// [`request::InitChain::validators`](super::request::InitChain::validators).
+    ///
+    /// If this list is nonempty, it will be the initial validator set, instead
+    /// of the one given in
+    /// [`request::InitChain::validators`](super::request::InitChain::validators).
+    pub validators: Vec<ValidatorUpdate>,
+    /// Initial application hash.
+    pub app_hash: Bytes,
+}
+
+/// Returns data queried from the application.
+///
+/// [ABCI documentation](https://docs.tendermint.com/master/spec/abci/abci.html#query)
+#[derive(Clone, PartialEq, Eq, Debug, Default)]
+pub struct Query {
+    /// The response code for the query.
+    pub code: u32,
+    /// The output of the application's logger.
+    ///
+    /// **May be non-deterministic**.
+    pub log: String,
+    /// Additional information.
+    ///
+    /// **May be non-deterministic**.
+    pub info: String,
+    /// The index of the key in the tree.
+    pub index: i64,
+    /// The key of the matching data.
+    pub key: Bytes,
+    /// The value of the matching data.
+    pub value: Bytes,
+    /// Serialized proof for the value data, if requested, to be verified against
+    /// the app hash for the given `height`.
+    pub proof: Option<merkle::Proof>,
+    /// The block height from which data was derived.
+    ///
+    /// Note that this is the height of the block containing the application's
+    /// Merkle root hash, which represents the state as it was after committing
+    /// the block at `height - 1`.
+    pub height: i64,
+    /// The namespace for the `code`.
+    pub codespace: String,
+}
+
+/// Returns events that occurred when beginning a new block.
+///
+/// [ABCI documentation](https://docs.tendermint.com/master/spec/abci/abci.html#beginblock)
+#[derive(Clone, PartialEq, Eq, Debug, Default)]
+pub struct BeginBlock {
+    /// Events that occurred while beginning the block.
+    pub events: Vec<Event>,
+}
+
+/// Returns the result of checking a transaction for mempool inclusion.
+///
+/// [ABCI documentation](https://docs.tendermint.com/master/spec/abci/abci.html#checktx)
+#[derive(Clone, PartialEq, Eq, Debug, Default)]
+pub struct CheckTx {
+    /// The response code.
+    ///
+    /// Transactions where `code != 0` will be rejected; these transactions will
+    /// not be broadcast to other nodes or included in a proposal block.
+    /// Tendermint attributes no other value to the response code.
+    pub code: u32,
```

Comment:
There's a struct called `Code` in `tendermint/src/abci/code.rs` that we could use here.

Reply:
Yes, but I didn't use it because I'm not sure that type is well-modeled: it's possible to create an error `Code` with value `0` that will be treated as a success by Tendermint, which doesn't seem right.  Maybe using a `NonZeroU32` would work, but I didn't want to get into it yet.

Reply:
Yes, you're right - we could leave this as a `u32` for now then :+1: